mount/mount_darwin.go

```diff
@@ -0,0 +1,42 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"fmt"
+	"os/exec"
+)
+
+// Mount to the provided target.
+func (m *Mount) mount(target string) error {
+	// See https://github.com/slonopotamus/containerd-darwin-mount-helper for reference implementation
+	const commandName = "containerd-darwin-mount-helper"
```

Comment:
Wondering if we can reach consensus on reviving this PR if this is just an environment variable like `CONTAINERD_DARWIN_MOUNT_HELPER` or `CONTAINERD_DARWIN_MOUNT_HELPER_EXPERIMENTAL` that takes the path of the binary that is compatible with https://github.com/slonopotamus/containerd-darwin-mount-helper .

cc @containerd/committers

Reply:
`CONTAINERD_DARWIN_MOUNT_HELPER_EXPERIMENTAL` sounds fairly reasonable to me.

Reply:
I am not a big fan of the generic helper interface which would be Darwin specific and not a well defined plugin interface. The idea around a large mount plugin interface doesn't really need to apply here either though, as the idea behind that interface before was always to go from custom mount types to system mount types. Even though "bindfs" is not a default util, I think we could easily just call out to "bindfs" and "hdiutil" in the Darwin mount code to cover a large range of the normal "system" mounts that would be expected there. Since "bindfs" is not a default, we could make that swappable or disabled via environment variables, but still utilizing the bindfs args. For apfs or other block type mounts, hdiutil can be relied on without needing to have custom binaries which just wrap hdiutil.

Reply:
@slonopotamus @mxpv WDYT? ↑

Reply:
Whoops, I somehow missed this.

1. Both containerd and buildkit expect across numerous places in codebase that they can just `umount()` things. bindfs working through either [macFUSE](https://osxfuse.github.io/) or [fuse-t](https://www.fuse-t.org/) supports that.
2. Containerd expects that bind-mounts exist and work. Even `native` snapshotter depends on bind mounts: https://github.com/containerd/containerd/blob/v2.0.0-rc.4/plugins/snapshots/native/native_default.go#L21, that's why https://github.com/darwin-containers/containerd/commit/416899fc8e81a80a4b09b59c801f98d36ddc0e74 was applied.
3. I agree with @dmcgowan that introducing a plugin API is an overkill. And it is very likely to be a weird plugin API: do we expect that a single helper is supposed to support all kinds of mounts? If there are multiple helpers, codebase needs to track which helper was used for mounting to use the same helper for unmount. This even can be across processes (containerd cleans up things in case shim dies).
4. Yes, binds is kind of a hack. Nobody says currently this is a production-scale code. If there emerges a better solution in the future, I will raise both of my hands for using it. But hacks like https://github.com/darwin-containers/containerd/commit/416899fc8e81a80a4b09b59c801f98d36ddc0e74 are even worse.
5. We're talking about a diff with _negative_ line delta :)

----

So, I'd like to split concerns.

First one is the need of bind-mounts that is not going to go anywhere. If you have other solutions besides `bindfs`, let's discuss them. if there are none, I don't see what's the benefit of adding wrapper layer. And API that is architected based on a single usecase usually fails to support the second when it emerges.

Second one is ineffectiveness of `native` snapshotter. This is a separate issue that can be solved independently (either via sparse-bundle approach suggested in #5935 or some FUSE solution or new [FSKit](https://developer.apple.com/documentation/fskit?language=objc) API). Even if `native` snapshotter is slow, it is functionally _correct_ and allows experimenting. My changes do not aim to solve this issue, I am only concentrated on bind-mounts that will be needed regardless of what snapshotter is used.

Reply:
We have failed to reach consensus on using a thirdparty FUSE for bindfs, but I guess we can easily reach consensus on using FSKit, as FSKit is Apple's official framework

Reply:
Unfortunately, FSKit is just an API that provides FUSE-like functionality. If someone manages to create an FS that implements bind-mounts on top of FSKit, that FS will still be a third-party product.

Reply:
Okay, I have an idea. I'll make another PR that only reorganizes files so it is easier to add darwin-specific bind-mount implementation, but won't actually add one. This will allow my containerd fork to have a smaller diff.

Reply:
See #10658.

Reply:
@AkihiroSuda Let's suppose _someone_ created a bindfs-like FOSS filesystem based on [FSKit](https://developer.apple.com/documentation/fskit). Would dependency on such thing be considered acceptable for containerd as a bind mounts implementation? According to FSKit documentation, it would be used via `mount -t MyAwesomeBind /src /dst`. Plain `unmount` would also work for this fs. Given that Darwin doesn't have builtin `bind` fs type, such implementation could even just use that name.

Reply:
SGTM, maybe it should be rather a built-in feature of containerd

cc @mxpv @dmcgowan

Reply:
> maybe it should be rather a built-in feature of containerd

Well, the issue is that FSKit API is currently only available for Swift and Objective-C.

Reply:
Could we just call out to `hdiutil` for now?

Reply:
> Could we just call out to `hdiutil` for now?

No, because we don't know what args to pass to `hdiutil` yet.